### PR TITLE
LingvoDSL: Utilize DSL4j dictionary loader/parser

### PR DESCRIPTION
**Pull request by miurahr**
DSL4j is real parser for LingvoDSL

Todos:

- [x] Handles `[lang id=1]` case

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

### CODE REVIEW

File: test/src/org/omegat/core/dictionaries/LingvoDSLTest.java [hunk 1]
```diff
@@ -86,28 +80,29 @@ public void testReadArticleRussian() throws Exception {
...
     @Test
     public void testReadArticleFontStyles() throws Exception {
-        LingvoDSLDict dict = (LingvoDSLDict) new LingvoDSL().loadDict(TEST_DICT, ENGLISH);
+        LingvoDSLDict dict = (LingvoDSLDict) new LingvoDSL().loadDict(TEST_DICT);
         String word = "italic";
         List<DictionaryEntry> result = dict.readArticles(word);
         assertFalse(result.isEmpty());
         assertEquals(word, result.get(0).getWord());
-        assertEquals("Here is an <span style='font-style: italic'>italic</span> <strong>word</strong>.\n",
+        assertEquals("<p>Here is an <span style='font-style: italic'>italic</span> <strong>word</strong>.</p>\n",
```
Comment (miurahr) [inline test/src/org/omegat/core/dictionaries/LingvoDSLTest.java:99]: Here is difference. When parser visit `[/m]`, we put `</p>` for `[m1]` - `[m9]` so we put `<p>` also for `[m]`

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 2]
```diff
@@ -64,161 +52,226 @@
...
+
+        LingvoDSLDict(File file) throws Exception {
+            data = DslDictionary.loadDictionary(file);
+            htmlVisitor = new HtmlVisitor(file.getParent());
+        }
+
+        @Override
+        public List<DictionaryEntry> readArticles(String word) {
+            return data.lookup(word).getEntries(htmlVisitor).stream()
+                    .map(e -> new DictionaryEntry(e.getKey(), e.getValue()))
```
Comment (miurahr) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:77]: Here is a place we can see how to use the parser.

- `DslDictionary#loadDictionary` is a builder method to load `dsl` file.
- `DslDictionary#lookup` returns `DslResult` object that hold result articles' AST.
- `DslResult#getEntries` returns `List<Entry<String, T>>` when pass `visitor<T>` object.

**Comment by miurahr** [comment]
Now ready for review.

File: src/org/omegat/core/dictionaries/LingvoDSL.java (hunk 2 above)
Comment (amake) [reply src/org/omegat/core/dictionaries/LingvoDSL.java:77]: Thanks.

Two minor thoughts:

1. As I understand it, Streams are significantly less efficient (make much more garbage) than simple `for` loops for this kind of trivial transformation. You might consider switching to a `for` loop if you're at all worried about that.
2. The only difference between `readArticles` and `readArticlesPredictive` is the method called on `this.data`. You could consolidate the transformation to `List<DictionaryEntry>`.

File: src/org/omegat/core/dictionaries/LingvoDSL.java [hunk 3]
```diff
@@ -64,161 +52,226 @@
...
+         *
+         * @return result.
+         */
         @Override
-        public List<DictionaryEntry> readArticlesPredictive(String word) {
-            return data.lookUpPredictive(word).stream().map(e -> new DictionaryEntry(e.getKey(), e.getValue()))
-                    .collect(Collectors.toList());
+        public String getObject() {
+            if (sb == null) {
+                return "";
```
Comment (amake) [inline src/org/omegat/core/dictionaries/LingvoDSL.java:240]: Just an idea: You can make calling `getObject` before `start` an error and throw an exception in that case.

**Review by amake — APPROVED** [review]
Thanks, this looks quite nice. I'm glad we now have a rigorous approach.

Just let me know what you think about my minor comments.

File: src/org/omegat/core/dictionaries/LingvoDSL.java (hunk 2 above)
Comment (miurahr) [reply src/org/omegat/core/dictionaries/LingvoDSL.java:77]: Thank you for the feedback. I'd like to try both ideas.

**Comment by amake** [comment]
Thanks, I've squashed to `master`.

**Pull request closed without merging**